### ISSUE TITLE
Dynamic RLS w/ Active Directory Security Group Memberships

### ISSUE THREAD

**Issue by JamesDBartlett3** [DataVolumeXYZ/DataVolumeXYZ#31]
If you want to learn more about RLS in Power BI, Reza Rad (of RADACAD.com) is the absolute best resource on that topic, and there's an entire section of his website devoted to it here: https://radacad.com/tag/row-level-security

There are lots of different ways to do RLS, but my personal favorite is [Dynamic RLS](https://radacad.com/dynamic-row-level-security-with-power-bi-made-simple) coupled with a daily import of [all security group memberships from the Active Directory connector](https://community.fabric.microsoft.com/t5/Desktop/PowerBI-to-Query-AD-Group-Memberships/td-p/1590094) in Power BI. This technique is a bit more complicated to setup, but the payoff is that you'll never have to modify the RLS settings again, even with significant changes in the organization structure, as long as those changes are reflected in Active Directory. 

YMMV, but here's how I set it up at my org (Des Moines University):

I started by finding the field in my model that I wanted to filter on (in this case, it's "CollegeCode" - see attached screenshot), ideally a field containing a string that would also be found in the names of AD groups that should be granted access to that data. In my org's data, there are three possible CollegeCode values, each one corresponding to a college within the university, and each college having multiple Programs within it, so I asked our sysadmins to create three new AD groups with names like this: `PowerBIReports_{WorkspaceName}_{CollegeCode}`. _Note: If your org has strict naming conventions for AD security groups, fields, and values, you may be able to use existing AD groups for this project. Otherwise you'll need to create an AD group for each possible value in the column to be filtered in your model._

Next, I used Power Query to filter the Active Directory security group memberships table (mine is called "_UserRLS") to only include rows where the AD group name starts with `PowerBIReports_{WorkspaceName}_`. Then, I split the AD group name column on the last underscore character and dropped every column besides CollegeCode and UserName. Then I connected the CollegeCode column in the __UserRLS table to the CollegeCode column ([via a bridge table](https://www.phdata.io/blog/how-to-create-a-bridge-table-in-power-bi/)) in the Program dimension, which in turn filters all of the fact tables in the model.

![image](https://github.com/DataVolumeXYZ/DataVolumeXYZ/assets/37491308/b5f0a6fe-1202-478d-8714-abe584d155a8)

With those relationships setup, I only needed a single RLS role in the model (mine is called "User"), and that role only needed to filter the table containing the AD group memberships (__UserRLS). Here's the DAX table filter for the User role:
`__UserRLS[UserName] IN {USERNAME(), USERPRINCIPALNAME()}`

Finally, I published the model to the Power BI service and assigned the User role to those same three AD groups that the sysadmins created earlier. And that's it! The only thing determining whether a user can see data related to a given CollegeCode is whether they are a member of the AD security group whose name ends with that same CollegeCode. For example, if the user needs to see data from CollegeCode "ABC" then they need to be a member of the `PowerBIReports_{WorkspaceName}_ABC` security group.

So now, whenever a user joins the organization or moves to a different department, as long as our sysadmins update those Active Directory security group memberships, that user will automatically gain/lose access to the data in Power BI, and I don't even have to _know about it_, much less _do anything_. 

**_TL;DR: Dynamic RLS paired with an Active Directory security group memberships table in your model requires more setup, but the payoff is that you'll most likely never have to touch it again._**
